app/controllers/chargeback_controller.rb

```diff
@@ -589,13 +743,44 @@ def cb_rate_set_form_vars
       "weekly"  => "Weekly",
       "monthly" => "Monthly"
     }
+    # creating a new set of rate details whose rate is 0
+    rate_details_compute_for_add = @sb[:rate_details].select { |k| k.rate == "0" && k.rate_type == "Compute" }
+    rate_details_storage_for_add = @sb[:rate_details].select { |k| k.rate == "0" && k.rate_type == "Storage" }
+
+    # if dont exist rate details not show the add rate detail form
+    if rate_details_compute_for_add.empty? && rate_details_storage_for_add.empty?
+      @edit[:new][:show_form] = false
+    else
+      @edit[:new][:show_form] = true
+      @edit[:new][:levels] = %w(Compute Storage)
+
+      if !session[:edit].nil? && !session[:edit][:new][:level].nil?
+        @edit[:new][:level] = session[:edit][:new][:level]
+        @edit[:new][:group] = session[:edit][:new][:group]
+        rate_details_for_add = @sb[:rate_details].select { |k| k.rate == "0" && k.rate_type == @edit[:new][:level] }
+        @edit[:new][:groups] = chargeback_details_groups(rate_details_for_add)
+        if @edit[:new][:groups].empty?
+          @edit[:new][:metrics] = []
+        elsif @edit[:new][:groups].value?(@edit[:new][:group])
+          @edit[:new][:metrics] = chargeback_details_metrics(rate_details_for_add, @edit[:new][:group])
+        else
+          @edit[:new][:metrics] = chargeback_details_metrics(rate_details_for_add, @edit[:new][:groups].first.second)
+        end
+      else
+        @edit[:new][:level] = "Compute"
+        rate_details_for_add = !rate_details_compute_for_add.empty? ? rate_details_compute_for_add : rate_details_storage_for_add
+        @edit[:new][:groups] = chargeback_details_groups(rate_details_for_add)
+        @edit[:new][:metrics] = chargeback_details_metrics(rate_details_for_add, @edit[:new][:groups].first.second)
```

Comment:
It seems like all of these are calculating the level, groups, and metrics... maybe you can have methods that return these things so this method just sets `@edit[:new][:level]`, groups and metrics based on those methods and doesn't have much logic.